Created new_keymap.py, python version of new_keymap.sh

orngshdw opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Saw some pull requests (#5771 & #5191) & an issue (#5776) going in the direction of re-writing scripts in Python so I thought to submit a Python3 version of the `new_keymap.sh` script.

### Usage
I tried to make the script execute & behave similar to `new_keyboard.sh`:
```
$ ./new_keymap.py
Keyboard Name: planck
New Username: username
username keymap directory created in: /mnt/d/git/qmk_firmware/keyboards/planck/keymaps/username

Compile a firmware file with your new keymap by typing:

make planck:username

from the qmk_firmware directory
```
### Debugging
I've also added debugging messages to the script that will print when executing the script with the debug flag. For example:
```
$ ./new_keymap.py --debug
DEBUG: Current directory is /mnt/d/git/qmk_firmware/util
DEBUG: Changed to directory /mnt/d/git/qmk_firmware
Keyboard Name: planck
New Username: username
DEBUG: kb_path = /mnt/d/git/qmk_firmware/keyboards/planck
keymap_path_default = /mnt/d/git/qmk_firmware/keyboards/planck/keymaps/default
keymap_path = /mnt/d/git/qmk_firmware/keyboards/planck/keymaps/username

username keymap directory created in: /mnt/d/git/qmk_firmware/keyboards/planck/keymaps/username

Compile a firmware file with your new keymap by typing:

make planck:username

from the qmk_firmware directory
```

This may be overkill for a simple script though. lmk your thoughts on this, can remove.

### Help text
Users can also `--help` for more information:
```
$ ./new_keymap.py --help
usage: ./new_keymap.py

Script to make a new keymap for a keyboard of your choosing. This script
automates the copying of the default keymap into your own keymap.

optional arguments:
  -h, --help  show this help message and exit
  --debug     specify to turn on debugging messages
```

Please feel free to comment/advise if I should do anything extra or correct anything.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* new file, `new_keymap.py` created in `qmk_firmware/util/` directory:
Will be a python3 version of `new_keymap.sh` file

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

fauxpark commented:
Thanks for the contribution!

IMO, the new keymap script should be written and function like the new keyboard script is/does now, for consistency. That way you don't need to look up what the arguments are to use it, just run it and it will ask you everything it needs to know.

I also think we shouldn't bother with Python 2 compatibility - it's about to reach EOL in about 6 months, and I believe the Python dependencies that have been added recently are 3 only.

drashna commented:
Also, all platforms that we're supporting also support Python 3, so that is what we should be targetting.

alloranges commented:
@fauxpark  & @drashna: Got it.  I changed the shebang to python3, updated the syntax, and refactored the new_keymap.py script a bit to make it more closer to how new_keyboard.sh behaves (now prompts the user and prints error messages similarly). 

Yall may recognize some parts of it from #5191 (...such as `print_error`). I guess you can say I got some inspiration from there :)

Feel free to suggest any more changes needed

orngshdw commented:
@fauxpark  & @drashna: Got it.  I changed the shebang to python3, updated the syntax, and refactored the new_keymap.py script a bit to make it more closer to how new_keyboard.sh behaves (now prompts the user and prints error messages similarly). 

Yall may recognize some parts of it from #5191 (...such as `print_error`). I guess you can say I got some inspiration from there :)

Feel free to suggest any more changes needed

fauxpark commented on util/new_keymap.py:
```diff
@@ -0,0 +1,133 @@
+from distutils.dir_util import copy_tree
+
+
+usage_message = "./new_keymap.py <keyboard_path> <username>\n" + \
```
This can be removed since the script no longer takes these arguments.

orngshdw replied:
oh right, good catch. Thank you

orngshdw commented on util/new_keymap.py:
```diff
@@ -0,0 +1,133 @@
+    Gets script arguments provided in terminal
+    """
+    # define usage and description
+    parser = argparse.ArgumentParser(description=__doc__, usage=usage_message)
```
Will remove usage part here

skullydazed commented:
Heya, thanks for submitting this. I haven't looked at it in depth yet but wanted to leave a couple quick notes.

Python in QMK is still very early, we have a lot to sort out still. We may change our mind about some of this before this PR is accepted. Please be patient if it seems like we contradict ourselves as time goes on. :) In particular we need to develop a style guide for docstrings.

We're targeting python 3.5, because that's the newest version shipped with the platforms we target. Specifically it seems you use f-strings here, which are not available in python 3.5.

I have a work in progress branch for the first big python feature I've been working on here, I'm not sure if everything here is final yet but it should give you an idea of the direction we're going in: https://github.com/qmk/qmk_firmware/compare/compile_json

orngshdw commented:
Thank you @skullydazed, your compile_json branch was really useful as a style guide. I made some updates reflecting what I saw.

In response to your comment on beginning Python implementation; I think that it's ok for things to be a bit slower at the start as you all at QMK define coding style/standards that'll best serve you + the user community.
I'm just glad to contribute to something I've used for so long. Look forward to hopefully collaborating more on similar stuff in the future.

fauxpark commented on util/new_keymap.py:
```diff
@@ -40,6 +40,27 @@ def set_logging(debug=None):
     logging.basicConfig(format="%(levelname)s: %(message)s")
 
 
+def cd_qmk_root(expected_files):
```
I'm not sure how useful this would be, considering it's somewhat difficult to run this from anywhere other than `qmk_firmware` and `util`.

For simplicity, new_keyboard.sh and my improved new_keymap.sh only check if we're in `util` - if so, `cd` up once - then check for the existence of the `quantum` directory, a reasonably good indicator we're in the right place. They don't check that the current directory is `qmk_firmware` because you can of course git clone the repo into a differently named folder. It probably doesn't need much more than that.

orngshdw replied:
@fauxpark : I was considering the scenario (maybe edge case?) where someone is at a keyboard's directory already and wanted to create a new keymap. For that scenario my current script will fail unless they changed to the project's root directory (or to util/), execute the script, then cd back to the created directory. Give so many steps for that possible scenario, I thought maybe I can reduce the steps by making the script smarter with the cd_qmk_root function I wrote.

Also I saw a find_dir function in @skullydazed 's compile_json branch and thought it was a good idea to keep checking upwards until we're in the right directory:
```def find_dir(keyboard):
    """Locate the correct directory for storing a keymap.
    """
    for directory in ['.', '..', '../..', '../../..', '../../../..', '../../../../..']:
        basepath = os.path.normpath(os.path.join('keyboards', keyboard, directory, 'keymaps'))
        if os.path.exists(basepath):
            return basepath```

If you feel those two thoughts above weren't convincing enough for the function, I can definitely revert :)

skullydazed replied:
Long term I think we'll have functionality like that in a wrapper script or something. Having code like that all over the place is not my idea of a good time. :)

The `find_dir(keyboard)` function is to locate a keymap dir. To locate the qmk_firmware directory I actually just get the directory for the script and move up one level: https://github.com/qmk/qmk_firmware/compare/compile_json#diff-8cc916123ff32a1a94cfdeb970c4190cR10

fauxpark replied:
I like the idea, but I'm just thinking that if you wanted to create a new keymap with this script from a keyboard directory, you'd have to type something like `./../../util/new_keymap.py`, which feels a tad unwieldy to me, and too unlikely to be done by most people to be worth caring about.

If QMK had a bunch more of these little scripts which the install script, say, added to the `PATH`, that would be a different story (and would require even more thorough sanity checking than this 😄).

orngshdw replied:
Points taken. Thank you. I'll do a revert later today

skullydazed commented:
It's been some time since we've looked at this PR, and the python stuff has matured considerably since. At this point I think the long-term path for this PR is to refactor it as `qmk new-keymap`, which will mean moving it to `lib/python/cli/new/keymap.py`, and reworking it to use the milc entrypoint and argument parsing.

@orngshdw is this still something you're interested in working on?

orngshdw commented:
Hi @drashna, will gladly continue to work on this. I really like how clean the milc framework looks.

Am thinking some initial steps to start can be to:
1. refactor using `lib/python/qmk/cli/compile.py` as an example file + incorporate in docs/coding_conventions_python.md pointers (as well as yapfiy), 
2. move `new_keymap.py` to its new location & re-name  as `lib/python/cli/new/keymap.py`, 
_Atm though in master I'm seeing `lib/python/qmk/cli/`, should I use this `cli` directory instead? Also you all like me to make the `new/` dir or is it part of branch I can merge into this feature branch?_
and
3. push for CR comments

Let me know what you think. And please feel free to add more steps or advise.

skullydazed commented:
That seems like a good general plan. Let me know if anything seems confusing or hard to work with as you write this- you'll be one of the first people besides me to actually use milc.

I do want you to create the new dir, in the future other "new" command like `new-keyboard` and `new-layout` can be created there. I may change this pattern in the future if it proves hard to work with, but for now this is what I'm trying out.

You are correct that `lib/python/qmk/cli/` is the right place, I have a bad habit of retyping paths instead of just copy and paste. :)

orngshdw commented on lib/python/qmk/path.py:
```diff
@@ -11,7 +15,7 @@ def keymap(keyboard):
             The name of the keyboard. Example: clueboard/66/rev3
     """
     for directory in ['.', '..', '../..', '../../..', '../../../..', '../../../../..']:
-        basepath = os.path.normpath(os.path.join('keyboards', keyboard, directory, 'keymaps'))
```
Felt that directory was meant to proceed other paths to join.

orngshdw commented on lib/python/qmk/path.py:
```diff
@@ -20,6 +24,56 @@ def keymap(keyboard):
     raise NoSuchKeyboardError('Could not find keymaps directory for: %s' % keyboard)
 
 
+def check_directories(*directories, exist=True, dont_raise=False):
```
Added this generic helper as something to use to check presence/absence of a directory.

orngshdw commented on lib/python/qmk/path.py:
```diff
@@ -20,6 +24,56 @@ def keymap(keyboard):
+    return False if error_paths else True
+
+
+def create_keymap_directory(default_keymap_directory, new_keymap_directory):
```
Looking back I think I will change this to something more general. Probably:  
```python
def copy_directory(existing_directory, new_directory):
```

orngshdw commented on lib/python/qmk/errors.py:
```diff
@@ -1,6 +1,17 @@
-class NoSuchKeyboardError(Exception):
-    """Raised when we can't find a keyboard/keymap directory.
+class DirectoryError(Exception):
```
I added a generically named error to raise directory type errors. I used this DirectoryError exception to raise an error to tell the user a path did/didn't exist.

May make more sense to change this to something more generic like `class QmkError(Exception):` and create a new error class that inherits it, like `class DirectoryError(QmkError):`

orngshdw commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
```
I rewrote the new_keymap.py file but scoped out setting logging level for this first draft. If you all believe the we can scope in setting the logging level for this file, see the todos (line 18) for the debug code

orngshdw commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,75 @@
+#   the "ModuleNotFoundError: No module named 'milc'" error
+#   am executing file from the lib/python/
+path = os.path.abspath(".")
+sys.path.append(path)
```
I wasn't sure how we'll be calling this script. Been calling it directly and added these `os.path.abspath` & `sys` commands so that I can test the script. Will need guidance here.

skullydazed replied:
I don't have time to look at everything but I did notice this- you can remove those lines and the shebang and then call this with `bin/qmk new-keymap`. That's the preferred way to run these.

orngshdw replied:
Removed lines so am resolving.

orngshdw commented:
Heys @skullydazed  Push a first draft. Took a stab at incorporating the milc framework. Let me know if it's the correct general directory you're expecting. And lmk if anything need to be changed up to be more clearer.

skullydazed commented:
Also, can you rebase and force push? You've pulled in all the changes from future. If you're not sure how to do that someone on discord can help walk you through it.

orngshdw commented:
Ok I posted in #help . I feel like I may need to reset back to before I merge in latest master and redo the changes..maybe something like:
```bash
# reset to commit before merging in latest master
git reset HEAD~2 --hard
git rebase master  # master will have latest commits
# 
# redo changes, add, commit, push
```
We'll see :)

orngshdw commented:
Did a `git rebase -i master` and dropped the squash commit that merged in latest master.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+
+
+# TODO (orngshdw): delete before merging to master
+#   for reference until we decide if we need logging/debugging for this script
```
FYI milc provides extensive logging and argument parsing already. Use `cli.log` to access the pre-configured [`logging`](https://docs.python.org/3.5/library/logging.html) objects. My next project is to revamp the configuration engine, but in the meantime you can set command line flags and access them via `cli.config.general` or `cli.args`, depending on whether it's a flag (--arg or -arg) or a positional argument.

I know this is probably holdover from the standalone script, but I wanted to provide that context until the full documentation is written.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+@cli.entrypoint('Create a new keymap.c from the default keyboard of your choosing')
+def main(cli):
+
+    # get keyboard
```
I'd like it if the user can provide these options as command line arguments, with a fallback to asking the user for any information not provided. This is not strictly necessary to merge this PR if you'd like to leave it to a future PR (by yourself or someone else.)

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+
+    # get keyboard
+    keyboard = input("Keyboard Name: ")
+    # get username
```
It would be good to refine your comments a bit. This one and the previous one could probably collected into a single comment that says "Collect information the user didn't provide".

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+    username = input("New Username: ")
+
+    # generate keymap paths
+    kb_path = os.path.abspath(keymap(keyboard))
```
os.path.abspath() is not necessary here. The current directory for python code is always the top level of `qmk_firmware`.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+    create_keymap_directory(keymap_path_default, keymap_path)
+
+    # end message to user
+    print("%s keymap directory created in: %s\n" % (
```
You should use `cli.log.info` here (and down below) instead of `print()`, and bring them up into 1 line.

orngshdw replied:
just for clarification, did you want all 4 print statements to be consolidated to one line or each of the print statements to be one line?

skullydazed replied:
In this case yes. Generally we want to split things like that into multiple lines when they get to be too complex, and these are all simple.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+
+    # check directories
+    check_directories(kb_path, keymap_path_default)
+    check_directories(keymap_path, exist=False)
```
```suggestion

    if os.path.exists(keymap_path):
        cli.log.error('Keymap %s already exists!', keymap_path)
        exit(1)
```

orngshdw replied:
Will we see the cli.log.error when running the script normally? If not is it ok to substitute `exit(1)` with `raise FileExistsError(msg)` or `FileNotFoundError` to communicate the error better?

skullydazed replied:
Yes, by default we're printing INFO and higher messages.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+    keymap_path = kb_path + "/" + username
+
+    # check directories
+    check_directories(kb_path, keymap_path_default)
```
```suggestion
    if not os.path.exists(kb_path):
        cli.log.error('Keyboard does not exist: %s', kb_path)
        exit(1)

    if not os.path.exists(keymap_path_default):
        cli.log.error('Default keymap for %s does not exist: %s', kb_path, keymap_path_default)
        exit(1)
```

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,65 @@
+    check_directories(keymap_path, exist=False)
+
+    # create user directory with default keymap files
+    create_keymap_directory(keymap_path_default, keymap_path)
```
```suggestion
    shutil.copytree(keymap_path_default, keymap_path, symlinks=True)
```

orngshdw commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,39 @@
+@cli.argument('-u', '--username', help='Specify any name for the new keymap directory')
+@cli.entrypoint('Create a new keymap the keyboard of your choosing')
+def main(cli):
+    # ask for user input if keyboard or username was not provided in the command line
```
@skullydazed I may have missed your comment about saying "Collect information the user didn't provide" when I was resolving comments from before. lmk if you still feel I can reword what I got here to be closer to what you suggested earlier for more clarity.

skullydazed replied:
You have the correct behavior here, I'm fine with this.

You probably want to put a docstring on main(), however, because pydoc will interpret the comment as documentation.

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,39 @@
+    # end message to user
+    cli.log.info("%s keymap directory created in: %s\n" % (username, keymap_path,) +
+                 "Compile a firmware file with your new keymap by typing: \n" +
+                 "make %s:%s\nfrom the qmk_firmware directory" % (keyboard, username,))
```
This could help promote the CLI better.

```suggestion
                 "qmk compile -kb %s -km %s", keyboard, username)
```

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,39 @@
+
+    # check directories
+    if not os.path.exists(kb_path):
+        cli.log.error('Keyboard %s does not exist!' % kb_path)
```
```suggestion
        cli.log.error('Keyboard %s does not exist!', kb_path)
```

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,39 @@
+        cli.log.error('Keyboard %s does not exist!' % kb_path)
+        exit(1)
+    if not os.path.exists(keymap_path_default):
+        cli.log.error('Keyboard default %s does not exist!' % keymap_path_default)
```
```suggestion
        cli.log.error('Keyboard default %s does not exist!', keymap_path_default)
```

skullydazed commented on lib/python/qmk/cli/new/keymap.py:
```diff
@@ -0,0 +1,39 @@
+    shutil.copytree(keymap_path_default, keymap_path, symlinks=True)
+
+    # end message to user
+    cli.log.info("%s keymap directory created in: %s\n" % (username, keymap_path,) +
```
I can't create a multiline suggestion but the vars should be parameters of cli.log.info here.

skullydazed commented:
Thanks!
